File: NNN-Solution/NNN/Components/Autodiff/TensorInitializations.cs
using NNN.Components.Utilities;

namespace NNN.Components.Autodiff;

public partial class Tensor
{
    // Initialization functions

    /// <summary>
    /// Creates a new tensor instance with the given dimensions.
    /// </summary>
    /// <param name="dims">Array containing the dimensions of the new tensor.</param>
    /// <param name="requiresGrad">Whether it is necessary to calculate the tensor's gradient.</param>
    public Tensor(int[] dims, bool requiresGrad = false)
    {
        // Calculate shape data
        Dimensions = (int[])dims.Clone();
        Strides = ComputeStrides(dims);

        // Calculate linear size
        int size = 1;
        foreach (var dim in dims) size *= dim;

        Data = new double[size]; // initialize linear array of values

        // Initialize linear array of gradients if necessary
        RequiresGrad = requiresGrad;
        if (RequiresGrad) Grad = new double[size];
    }

    /// <summary>
    /// Parameterless constructor for Json serializer.
    /// </summary>
    public Tensor() { }

    /// <summary>
    /// Initializes a new weights tensor using He Initialization.
    /// </summary>
    /// <param name="inputCount">Number of inputs to the layer.</param>
    /// <param name="neuronCount">Number of neurons in the layer.</param>
    /// <returns>Weights tensor with values initialized using He Initialization.</returns>
    public static Tensor InitWeights(int inputCount, int neuronCount)
    {
        Tensor weights = new([inputCount, neuronCount], true);

        double stdDev = Math.Sqrt(2.0 / inputCount);
        for (int i = 0; i < weights.ElementCount; i++)
        {
            weights[i] = MathUtils.NextGaussian(0, stdDev);
        }

        return weights;
    }

    /// <summary>
    /// Initializes a new bias tensor with non-zero values.
    /// </summary>
    /// <param name="neuronCount">Number of neurons in the layer.</param>
    /// <returns>Bias tensor with non-zero values.</returns>
    public static Tensor InitBiases(int neuronCount) => Scalar(0.01, [neuronCount], true);

    public static Tensor InitKernels(int filterCount, int[] kernelDims, int inputChannels)
    {
        var dims = new int[kernelDims.Length + 2];
        dims[0] = filterCount;
        Array.Copy(kernelDims, 0, dims, 1, kernelDims.Length);
        dims[^1] = inputChannels;

        int fanIn = inputChannels;
        foreach (var dim in kernelDims)
        {
            fanIn *= dim;
        }

        Tensor kernels = new(dims, true);

        double stdDev = Math.Sqrt(2.0 / fanIn);
        for (int i = 0; i < kernels.ElementCount; i++)
        {
            kernels[i] = MathUtils.NextGaussian(0, stdDev);
        }

        return kernels;
    }

    /// <summary>
    /// Creates a tensor instance filled with a single scalar value.
    /// </summary>
    /// <param name="value">Scalar value to fill the new tensor with.</param>
    /// <param name="dims">Dimensions of the new tensor.</param>
    /// <param name="requiresGrad">Whether it is necessary to calculate the gradient of the new tensor.</param>
    /// <returns>New tensor instance filled with the given scalar value.</returns>
    public static Tensor Scalar(double value, int[] dims, bool requiresGrad = false)
    {
        Tensor t = new(dims, requiresGrad);
        Array.Fill(t.Data, value);
        return t;
    }

    /// <summary>
    /// Creates a deep copy of the current tensor.
    /// </summary>
    /// <returns>Deep copy tensor detached from the existing autograd graph.</returns>
    public Tensor Copy()
    {
        Tensor copy = new(Dimensions, false);
        Array.Copy(Data, copy.Data, ElementCount);
        return copy;
    }
}

File: NNN-Solution/NNN/TensorInitializations.cpp
#include "pch.h"
#include "Tensor.h"
#include "MathUtils.h"

/* Initialization */

// Computes the strides of the tensor using the given dimensions.
std::vector<int> Tensor::compute_strides(const std::vector<int>& dims)
{
	int n = (int)dims.size();
	std::vector<int> strides(n);

	strides[n - 1] = 1;
	for (int i = n - 2; i >= 0; i--)
	{
		strides[i] = strides[i + 1] * dims[i + 1];
	}

	return strides;
}

Tensor::Tensor() {}

Tensor::Tensor(const std::vector<int> dims, bool req_grad) : _dimensions(dims), requires_grad(req_grad)
{
	_strides = compute_strides(dims);

	// Compute linear size and initialize data and gradient vectors

	int size = 1;
	for (int dim : dims)
	{
		size *= dim;
	}

	_data.assign(size, 0.0);
	
	if (requires_grad)
	{
		_grad.assign(size, 0.0);
	}
}

Tensor::Tensor(double value, const std::vector<int> dims, bool req_grad) : _dimensions(dims), requires_grad(req_grad)
{
	_strides = compute_strides(dims);

	// Compute linear size and initialize data and gradient vectors

	int size = 1;
	for (int dim : dims)
	{
		size *= dim;
	}

	_data.assign(size, value);

	if (requires_grad)
	{
		_grad.assign(size, 0.0);
	}
}

std::shared_ptr<Tensor> Tensor::init_weights(int input_count, int neuron_count)
{
	auto weights = std::make_shared<Tensor>(std::vector<int>{input_count, neuron_count}, true);

	// Initialize weight values using He initialization
	double std_dev = std::sqrt(2.0 / input_count);
	const int element_count = weights->element_count();
	for (int i = 0; i < element_count; ++i)
	{
		weights->_data[i] = MathUtils::next_gaussian(0.0, std_dev);
	}

	return weights;
}

std::shared_ptr<Tensor> Tensor::init_biases(int neuron_count)
{
	return std::make_shared<Tensor>(0.01, std::vector<int>{neuron_count}, true);
}

std::shared_ptr<Tensor> Tensor::init_kernels(int filter_count, const std::vector<int>& kernel_dims, int input_channels)
{
	// Compute kernels tensor dimensions
	std::vector<int> dims;
	dims.reserve(kernel_dims.size() + 2);
	dims.push_back(filter_count);
	int fan_in = input_channels;
	for (int dim : kernel_dims)
	{
		dims.push_back(dim);
		fan_in *= dim;
	}
	dims.push_back(input_channels);

	auto kernels = std::make_shared<Tensor>(dims, true);

	// Initialize kernel values using He initialization
	double std_dev = std::sqrt(2.0 / fan_in);
	const int element_count = kernels->element_count();
	for (int i = 0; i < element_count; ++i)
	{
		kernels->_data[i] = MathUtils::next_gaussian(0.0, std_dev);
	}

	return kernels;
}

std::shared_ptr<Tensor> Tensor::copy() const
{
	auto copy = std::make_shared<Tensor>(_dimensions, requires_grad);
	copy->_data = _data;
	return copy;
}